File: Medium/1249.cs
using System.Collections.Generic;
using System.Text;

public class Solution {
  public string MinRemoveToMakeValid(string s) {
    Stack<int> stack = new Stack<int>();
    HashSet<int> removeSet = new HashSet<int>();

    for (int i = 0; i < s.Length; i++) {
      if (s[i] == '(') {
        stack.Push(i);
      }
      else if (s[i] == ')') {
        if (stack.Count > 0) {
          stack.Pop();
        }
        else {
          // 没有左括号和他匹配，removeSet 添加这个 index
          removeSet.Add(i);
        }
      }
    }

    while (stack.Count > 0) {
      // 填加“没有右括号和它匹配“的左括号的 index
      removeSet.Add(stack.Pop());
    }

    StringBuilder result = new StringBuilder();

    for (int i = 0; i < s.Length; i++) {
      if (!removeSet.Contains(i)) {
        result.Append(s[i]);
      }
    }

    return result.ToString();
  }
}

File: Medium/1249.py
class Solution:
  def minRemoveToMakeValid(self, s: str) -> str:
    bracket_index = []
    invalid_index = [False] * len(s)
    result = []

    for i in range(len(s)):
      # 如果遇到左括号
      if s[i] == "(":
        bracket_index.append(i)
        invalid_index[i] = True

      # 如果遇到右括号
      if s[i] == ")":
        if len(bracket_index) == 0:
          # 这个 ")" 的前面不会有 "(" 和它匹配
          invalid_index[i] = True
        else:
          # 说明找到匹配的了
          invalid_index[bracket_index.pop()] = False

    for i in range(len(s)):
      if not invalid_index[i]:
        result.append(s[i])

    return "".join(result)